pkg/wireguard/operator/operator.go

```diff
@@ -0,0 +1,369 @@
+		ipv6Alloc, err = ipallocator.NewCIDRRange(subnetV6)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &Operator{
+		restoring:             true,
+		ciliumNodeUpdater:     ciliumNodeUpdater,
+		ipv4Alloc:             ipv4Alloc,
+		ipv4ByNode:            ipv4ByNode,
+		ipv4AllocAfterRestore: ipv4AllocAfterRestore,
+		ipv6Alloc:             ipv6Alloc,
+		ipv6ByNode:            ipv6ByNode,
+		ipv6AllocAfterRestore: ipv6AllocAfterRestore,
+	}, nil
+}
+
+// AddNode is called by cilium-operator when a new CiliumNode is created.
+func (o *Operator) AddNode(n *v2.CiliumNode) error {
+	o.Lock()
+	defer o.Unlock()
+
+	if option.Config.EnableIPv4 {
+		if err := o.allocateIP(n, types.IPv4); err != nil {
+			return err
+		}
+	}
+
+	if option.Config.EnableIPv6 {
+		if err := o.allocateIP(n, types.IPv6); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// UpdateNode is called by cilium-operator when an existing CiliumNode is updated.
+func (o *Operator) UpdateNode(n *v2.CiliumNode) error {
+	o.Lock()
+	defer o.Unlock()
+
+	if option.Config.EnableIPv4 {
+		if err := o.allocateIP(n, types.IPv4); err != nil {
+			return err
+		}
+	}
+
+	if option.Config.EnableIPv6 {
+		if err := o.allocateIP(n, types.IPv6); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DeleteNode is called by cilium-operator when an existing CiliumNode is removed.
+func (o *Operator) DeleteNode(n *v2.CiliumNode) {
+	o.Lock()
+	defer o.Unlock()
+
+	nodeName := n.ObjectMeta.Name
+
+	if o.restoring {
+		log.WithField(logfields.NodeName, nodeName).Warn("Received node delete while restoring")
+	}
+
+	if option.Config.EnableIPv4 {
+		o.releaseIP(n, nodeName, types.IPv4)
+	}
+
+	if option.Config.EnableIPv6 {
+		o.releaseIP(n, nodeName, types.IPv6)
+	}
+}
+
+func (o *Operator) releaseIP(n *v2.CiliumNode, nodeName string, family types.Family) {
+	var (
+		ipAlloc  *ipallocator.Range
+		ipByNode map[string]net.IP
+	)
+
+	switch family {
+	case types.IPv4:
+		ipAlloc = o.ipv4Alloc
+		ipByNode = o.ipv4ByNode
+	case types.IPv6:
+		ipAlloc = o.ipv6Alloc
+		ipByNode = o.ipv6ByNode
+	default:
+		panic(fmt.Sprintf("unsupported family: %s", family))
+	}
+
+	ip, found := findWireguardIP(n, family)
+
+	if !found {
+		// Maybe cilium-agent has removed the IP addr from CiliumNode, so fallback
+		// to local cache to determine the IP addr.
+		ip, found = ipByNode[nodeName]
+	}
+
+	if found {
+		ipAlloc.Release(ip)
+		delete(ipByNode, nodeName)
+
+		log.WithFields(logrus.Fields{
+			logfields.NodeName: nodeName,
+			logfields.IPAddr:   ip,
+		}).Info("Released wireguard IP")
+	}
+}
+
+// RestoreFinished is called by cilium-operator once it has received all
+// CiliumNodes from kube-apiserver during its restore.
+//
+// RestoreFinished removes IP allocations of CiliumNodes which no longer exist.
+func (o *Operator) RestoreFinished() error {
+	o.Lock()
+	defer o.Unlock()
+
+	for nodeName := range o.ipv4AllocAfterRestore {
+		ip, err := o.ipv4Alloc.AllocateNext()
```

Comment:
Seems to me that the entries of `ipv4AllocAfterRestore` are never deleted. This is probably fine since AFAICT, we can never go from the restoring state to the non-restoring state, but maybe is worth removing the entries anyway?

Reply:
Good catch! We'll clear it to allow for GC.